A105-max_concurrent_streams-connection-scaling.md

```diff
@@ -0,0 +1,747 @@
+def RetryRpcsFromQueue(self):
+  while len(self.queue) > 0:
+    connection = self.ChooseConnection()
+    # Stop at the first RPC that gets queued.
+    if connection is None:
+      break
+    # Otherwise, send the RPC on the connection.
+    connection.SendRpc(self.queue.front())
+    self.queue.pop()
+
+# Starts a new connection attempt.
+def StartConnectionAttempt(self):
+  if self.backoff_state is None:
+    self.backoff_state = BackoffState()
+  self.connection_attempt = ConnectionAttempt()
+
+# Called when a connection attempt succeeds.
+def OnConnectionAttemptSucceeded(self, new_connection):
+  self.connections.append(new_connection)
+  self.pending_backoff_timer = None
+  self.backoff_state = None
+  self.RetryRpcsFromQueue()
+
+# Called when a connection attempt fails.  This puts us in backoff.
+def OnConnectionAttemptFailed(self):
+  self.connection_attempt = None
+  self.pending_backoff_timer = Timer(self.backoff_state.NextBackoffDelay())
+
+# Called when the backoff timer fires.  Will trigger a new connection
+# attempt if there are RPCs in the queue.
+def OnBackoffTimer(self):
+  self.pending_backoff_timer = None
+  self.RetryRpcsFromQueue()
+
+# Called when an established connection fails.  Will trigger a new
+# connection attempt if there are RPCs in the queue.
+def OnConnectionFailed(self, failed_connection):
+  self.connections.remove(failed_connection)
+  if len(self.connections) == 0:
+    for rpc in self.queue:
+      # RPC will be eligible for transparent retries.
+      rpc.Fail(UNAVAILABLE)
+    self.queue = []
+  else:
+    self.RetryRpcsFromQueue()  # Maybe trigger new connection attempt.
+
+# Called when a connection reports a new MAX_CONCURRENT_STREAMS value.
+# May send RPCs on the connection if there are any queued and the
+# MAX_CONCURRENT_STREAMS value has increased.
+def OnConnectionReportsNewMaxConcurrentStreams(
+    self, connection, max_concurrent_streams):
+  connection.max_concurrent_streams = max_concurrent_streams
+  self.RetryRpcsFromQueue()
+
+# Called when an RPC completes on one of the subchannel's connections.
+def OnRpcComplete(self, connection):
+  connection.rpcs_in_flight -= 1
+  self.RetryRpcsFromQueue()
+
+# Called when the max_connections_per_subchannel value changes.
+def OnMaxConnectionsPerSubchannelChanged(self, max_connections_per_subchannel):
+  self.max_connections_per_subchannel = max_connections_per_subchannel
+  self.RetryRpcsFromQueue()
+```
+
+### PickFirst Changes
+
+As mentioned above, the pick_first LB policy will need to handle two new
+state transitions from subchannels.  Previously, a subchannel in READY
+state could transition only to IDLE state; now it will be possible for
+the subchannel to instead transition to CONNECTING or TRANSIENT_FAILURE
+states.
+
+If the selected (READY) subchannel transitions to CONNECTING or
+TRANSIENT_FAILURE state, then pick_first will go back into CONNECTING
+state.  It will start the happy eyeballs pass across all subchannels,
```

Comment:
Do we want to say it will trigger a name resolution refresh, or is it clear because PF left READY?